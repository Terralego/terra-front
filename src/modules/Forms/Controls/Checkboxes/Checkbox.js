import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox as BPCheckbox } from '@blueprintjs/core';

const Checkbox = ({ checked, label, onToggle, value }) => (
  <BPCheckbox
    onChange={() => onToggle(value)}
    label={label || value}
    checked={checked}
  />
);

Checkbox.propTypes = {
  value: PropTypes.string,
  label: PropTypes.string,
  checked: PropTypes.bool,
  onToggle: PropTypes.func,
};

Checkbox.defaultProps = {
  value: '',
  label: '',
  checked: false,
  onToggle () {},
};

export default Checkbox;
