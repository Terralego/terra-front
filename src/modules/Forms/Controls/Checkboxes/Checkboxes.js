import React from 'react';
import PropTypes from 'prop-types';
import { Spinner } from '@blueprintjs/core';

import Checkbox from './Checkbox';
import formatValues from '../formatValues';
import NoValues from '../NoValues';
import '../index.scss';

export class Checkboxes extends React.Component {
  static getDerivedStateFromProps ({ values }) {
    if (!values) return null;

    return {
      values: formatValues(values),
    };
  }

  state = {}

  onToggle = toggledValue => {
    const { value, onChange } = this.props;
    const newValue = value.includes(toggledValue)
      ? [...value.filter(val => val !== toggledValue)]
      : [...value, toggledValue];
    onChange(newValue);
  }

  render () {
    const { value: mainValue, label: mainLabel, loading } = this.props;
    const { values } = this.state;
    const { onToggle } = this;

    return (
      <div className="control-container">
        <p className="control-label">{mainLabel}</p>
        {loading && <Spinner size={20} />}
        {!loading && !values.length && <NoValues />}
        {!loading && values.map(({ label, value }) => (
          <Checkbox
            key={`${label}${value}`}
            onToggle={onToggle}
            value={value}
            label={label}
            checked={mainValue.includes(value)}
          />
        ))}
      </div>
    );
  }
}

Checkboxes.propTypes = {
  label: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.string]),
  // Prop is used in getDerivedStateFromProps
  // eslint-disable-next-line react/no-unused-prop-types
  values: PropTypes.arrayOf(PropTypes.string),
  onChange: PropTypes.func,
  loading: PropTypes.bool,
};

Checkboxes.defaultProps = {
  label: '',
  value: [],
  values: [],
  onChange () {},
  loading: false,
};

export default Checkboxes;
