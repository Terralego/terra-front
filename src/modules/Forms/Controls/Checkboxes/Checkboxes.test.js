import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Checkboxes from './Checkboxes';

jest.mock('@blueprintjs/core', () => ({
  Checkbox: () => <p>Checkbox</p>,
  Spinner: () => <p>Spinner</p>,
}));

it('should render correctly', () => {
  const tree = renderer.create((
    <Checkboxes
      label="Pwout"
      onChange={() => null}
      values={['pwet', 'wxd']}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render loading correctly', () => {
  const tree = renderer.create((
    <Checkboxes label="Pwout" loading />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should mount & update correctly', () => {
  const onChange = jest.fn();
  const wrapper = shallow((
    <Checkboxes
      label="Pwout"
      onChange={onChange}
      values={['pwet', 'wxd']}
      value={['foo']}
    />
  ));

  wrapper.instance().onToggle('pwet');
  expect(onChange).toHaveBeenCalledWith(['foo', 'pwet']);
  wrapper.instance().onToggle('foo');
  expect(onChange).toHaveBeenCalledWith([]);
});

it('should update values', () => {
  expect(Checkboxes.getDerivedStateFromProps({ })).toBe(null);
  expect(Checkboxes.getDerivedStateFromProps({ values: ['foo', 'bar'] })).toEqual({
    values: [{
      value: 'foo',
      label: 'foo',
    }, {
      value: 'bar',
      label: 'bar',
    }],
  });
});
