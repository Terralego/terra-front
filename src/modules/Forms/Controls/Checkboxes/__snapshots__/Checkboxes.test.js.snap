// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="control-container"
>
  <p
    className="control-label"
  >
    Pwout
  </p>
  <p>
    Checkbox
  </p>
  <p>
    Checkbox
  </p>
</div>
`;

exports[`should render loading correctly 1`] = `
<div
  className="control-container"
>
  <p
    className="control-label"
  >
    Pwout
  </p>
  <p>
    Spinner
  </p>
</div>
`;
